Models

nguyenmanh96 opened the pull request.

Commit:
homepage <redirect> login

Commit:
created relationship (80%)

nguyenmanh96 committed:
create repositories, fix data type of users table

DuongXuanBinh commented on app/Http/Controllers/Admin/AdminDashboardController.php:
```diff
@@ -4,6 +4,7 @@
 
 use App\Http\Controllers\Controller;
 use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
```
Cái này dùng ở đâu? 
Tại sao lại cần dùng DB ở đây?

nguyenmanh96 replied:
added by mistake, fixed.

DuongXuanBinh commented on app/Models/Department.php:
```diff
@@ -0,0 +1,20 @@
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Department extends Model
```
Toàn bộ model chưa hề có softdelete

nguyenmanh96 replied:
added $table->softdelete() into all tables.

DuongXuanBinh commented on app/Models/Result.php:
```diff
@@ -0,0 +1,20 @@
+
+class Result extends Model
+{
+    protected $table = 'results';
```
Không thấy có fillable.
Đọc cái này để biết tại sao cần fillable
https://viblo.asia/p/tim-hieu-mass-assignment-trong-laravel-eloquent-orm-bJzKmrbYZ9N

nguyenmanh96 replied:
Fixed all tables.

DuongXuanBinh commented on app/Models/RegisterSubject.php:
```diff
@@ -0,0 +1,13 @@
+{
+    protected $table = 'register_subjects';
+
+
```
Model này không có relation à?

nguyenmanh96 replied:
added.

DuongXuanBinh commented on app/Models/Result.php:
```diff
@@ -0,0 +1,20 @@
+    protected $table = 'results';
+
+
+    public function students(){
```
Cả 2 relation đều sai

nguyenmanh96 replied:
fixed.

DuongXuanBinh replied:
relation ở đây là dạng gì?

nguyenmanh96 replied:
nhuều nhiều, đã fix.

DuongXuanBinh commented on app/Models/Student.php:
```diff
@@ -0,0 +1,33 @@
+
+    public function subjects()
+    {
+        return $this->hasManyThrough(Subject::class, Result::class);
```
Nếu sử dụng hasManyThrough để lấy ra subjects như này là sai

nguyenmanh96 replied:
fixed.

DuongXuanBinh commented on app/Models/User.php:
```diff
@@ -3,43 +3,55 @@
 namespace App\Models;
 
 // use Illuminate\Contracts\Auth\MustVerifyEmail;
```
Xóa comment

nguyenmanh96 replied:
deleted.

DuongXuanBinh commented on app/Models/User.php:
```diff
@@ -3,43 +3,55 @@
-        'password',
-        'remember_token',
-    ];
+    public function getAuthIdentifierName()
```
Các hàm này ở đâu thế?

nguyenmanh96 replied:
When attaching the Authenticatable implementation to the model, the system suggests adding method stubs. Removed.

DuongXuanBinh commented on app/Providers/AppServiceProvider.php:
```diff
@@ -11,7 +23,12 @@ class AppServiceProvider extends ServiceProvider
     public function register(): void
     {
-        //
+        $this->app->bind(UserRepositoryInterface::class,UserRepository::class);
```
Mục đích của bind ở đây để làm gì?

nguyenmanh96 replied:
Means binding an interface UserRepositoryInterface to a specific implementation class UserRepository.
whenever something requests an instance of RepositoryInterface, it should provide an instance of BaseRepository.

DuongXuanBinh commented on app/Repositories/Repository/DepartmentRepository.php:
```diff
@@ -0,0 +1,21 @@
+        parent::__construct($department);
+    }
+
+    public function paginate($perPage = 10)
```
tại sao có paginate ở đây?

nguyenmanh96 replied:
Laravel Exercise
....................
-Department list needs pagination.
......................

DuongXuanBinh commented on database/migrations/2024_04_23_092705_create_users_table.php:
```diff
@@ -17,8 +17,8 @@ public function up(): void
             $table->string('avatar',255);
-            $table->string('google_token',50);
-            $table->unsignedBigInteger('student_id');
+            $table->string('google_token',50)->nullable();
```
Không được phép sửa migration. Nếu làm trong 1 team sẽ bị ảnh hưởng toàn bộ hệ thống.
Nếu muốn thay đổi cần tạo 1 migration mới để sửa

nguyenmanh96 replied:
https://www.youtube.com/watch?v=u86VdOc3asA)](https://www.youtube.com/watch?v=u86VdOc3asA

DuongXuanBinh commented on database/seeders/UsersTableSeeder.php:
```diff
@@ -0,0 +1,23 @@
+            'email' => 'admin@gmail.com',
+            'password' => bcrypt('12345678'),
+            'type' => 'Admin',
+            'avatar' => 'aaaa',
```
Avatar để dữ liệu default để còn hiển thị được ảnh. 
Tìm 1 ảnh giống như khi chưa set avatar của facebook cho vào source code, rồi đưa đường dẫn file vào làm giá trị mặc định của cột avatar trong DB

nguyenmanh96 replied:
Done.

nguyenmanh96 committed:
fix bug:
- add $table->softdelete() into all tables.
- create $fillable for all tables.
- fix relation.
- delete comments,Authenticatable method

DuongXuanBinh commented on app/Http/Controllers/Login/AuthController.php:
```diff
@@ -0,0 +1,31 @@
+            'email'=> $email,
+            'password'=> $password,
+       ])) {
+            $user = User::where('type','==','Admin')->get()->first();
```
Trong controller không dùng model

DuongXuanBinh replied:
submit login xong lấy user rồi xử lý gì?

nguyenmanh96 replied:
đã fix.

DuongXuanBinh commented on app/Models/RegisterSubject.php:
```diff
@@ -0,0 +1,26 @@
+        'status',
+    ];
+
+    public function student()
```
relation ở đây là gì? tên 2 relation chưa thống nhất

nguyenmanh96 replied:
relation nhiều nhiều. đã fix.

DuongXuanBinh commented on database/migrations/2024_04_23_092705_create_departments_table.php:
```diff
@@ -6,21 +6,18 @@
             $table->id();
             $table->string('name',100);
             $table->timestamps();
+            $table->softDeletes();
```
softDelete phải dùng trong cả model

DuongXuanBinh replied:
Không sửa migration

DuongXuanBinh commented on config/auth.php:
```diff
@@ -91,6 +91,7 @@
     */
 
     'passwords' => [
+        'algorithm'=> 'bcrypt',
```
Ko thêm cái này không login được à?

nguyenmanh96 replied:
vẫn có thể đăng nhập. đã xóa.

DuongXuanBinh commented on routes/web.php:
```diff
@@ -1,19 +1,30 @@
+    Route::get('/delete-user',function (){
+        $user = \App\Models\User::find(2);
+        $user->delete();
+        echo 'Da xoa item';
```
Không dùng tiếng Việt

nguyenmanh96 replied:
đã fix.

DuongXuanBinh commented on database/migrations/2024_04_29_022801_after_users_add_method.php:
```diff
@@ -0,0 +1,25 @@
+    public function down(): void
+    {
+        schema::table('users',function (Blueprint $table){
+            $table->string('avatar',255)->default('public.default_avt.png')->change();
```
tại sao đường dẫn lại là .?
chưa thấy có file ảnh trong source

nguyenmanh96 replied:
khi link ảnh đến ảnh lưu ở public ảnh vị vỡ nên e dẫn link web tạm cho có UI để sau sửa.
